internal/app/go-filecoin/internal/submodule/storage_miner_node.go

```diff
@@ -0,0 +1,508 @@
+// Returns false if this handler is no longer valid
+func (l heightThresholdListener) handle(ctx context.Context, chain []block.TipSet, sampler ChainSampler) (bool, error) {
+	if len(chain) < 1 {
+		return true, nil
+	}
+
+	h, err := chain[0].Height()
+	if err != nil {
+		return true, err
+	}
+
+	// check if we've hit finality and should stop listening
+	if h >= l.target+consensus.FinalityEpochs {
+		return false, nil
+	}
+
+	lcaHeight, err := chain[len(chain)-1].Height()
+	if err != nil {
+		return true, err
+	}
+
+	// if we have already seen a target tipset
+	if l.targetHit {
+		// if we've completely reverted
+		if h < l.target {
+			l.invalidCh <- struct{}{}
+			l.targetHit = false
+			// if we've re-orged to a point before the target
+		} else if lcaHeight < l.target {
+			l.invalidCh <- struct{}{}
+			err := l.sendRandomness(ctx, chain, sampler)
+			if err != nil {
+				return true, err
+			}
+		}
+		return true, nil
+	}
+
+	// otherwise send randomness if we've hit the height
+	if h >= l.target {
+		l.targetHit = true
+		err := l.sendRandomness(ctx, chain, sampler)
+		if err != nil {
+			return true, err
+		}
+	}
+	return true, nil
+}
+
+func (l heightThresholdListener) sendRandomness(ctx context.Context, chain []block.TipSet, sampler ChainSampler) error {
+	// assume chain not empty and first tipset height greater than target
+	firstTargetTipset := chain[0]
+	for _, ts := range chain {
+		h, err := ts.Height()
+		if err != nil {
+			return err
+		}
+
+		if h < l.target {
+			break
+		}
+		firstTargetTipset = ts
+	}
+
+	tsHeight, err := firstTargetTipset.Height()
+	if err != nil {
+		return err
+	}
+
+	randomness, err := sampler(ctx, types.NewBlockHeight(tsHeight))
+	if err != nil {
+		return err
+	}
+
+	l.seedCh <- storage.SealSeed{
+		BlockHeight: tsHeight,
+		TicketBytes: randomness,
+	}
+	return nil
+}
+
+// StorageMinerNodeAdapter is a struct which satisfies the go-storage-miner
+// needs of "the node," e.g. interacting with the blockchain, persisting sector
+// states to disk, and so forth.
+type StorageMinerNodeAdapter struct {
+	minerAddr  address.Address
+	workerAddr address.Address
+
+	newListener     chan heightThresholdListener
+	heightListeners []heightThresholdListener
+	listenerDone    chan struct{}
+
+	chain     *ChainSubmodule
+	messaging *MessagingSubmodule
+	msgWaiter *msg.Waiter
+	wallet    *WalletSubmodule
+}
+
+var _ storage.NodeAPI = new(StorageMinerNodeAdapter)
+
+// NewStorageMinerNodeAdapter produces a StorageMinerNodeAdapter, which adapts
+// types in this codebase to the interface representing "the node" which is
+// expected by the go-storage-miner project.
+func NewStorageMinerNodeAdapter(minerAddress address.Address, workerAddress address.Address, c *ChainSubmodule, m *MessagingSubmodule, mw *msg.Waiter, w *WalletSubmodule) *StorageMinerNodeAdapter {
+	return &StorageMinerNodeAdapter{
+		minerAddr:    minerAddress,
+		workerAddr:   workerAddress,
```

Comment:
> Because the miner node adapter does not track changes to worker address this means the miner node adapter will be broken when the miner changes their worker addr on chain.

@laser to investigate plumbing a worker-address-getting function (or channel?) to the go-storage-miner.

Reply:
don't feel like you have to for this PR.  I filed an issue so the work can be offloaded to another time or another person

Reply:
Thanks for noting this in an issue.